src/OpenCV/Core/Types/Mat.hs

```diff
@@ -242,3 +262,87 @@ matCopyToM dstM (V2 x y) src mbSrcMask = ExceptT $
+instance (p x, All p xs) => All p (x ': xs)
+
+class IsStatic (ds :: DS a)
+instance IsStatic ('S a)
+
+
+type Pixel = Vector
+type Steps = [Int]
+type Position = [Int]
+
+
+-- |Zips a given list of matrices of equal shape, channels, and depth, by
+-- applying the given function to the corresponding matrix elements at each
+-- position and storing the result at the corresponding position in a newly
+-- created matrix of the same shape, number of channels, and depth.
+zipMatWith :: forall (shape :: [DS Nat]) (channels :: Nat) (depth :: *)
+         . ( Storable depth, All IsStatic shape )
+        => ([Pixel depth] -> Pixel depth)
+        -> [Mat ('S shape) ('S channels) ('S depth)]
+        -> Maybe (Mat ('S shape) ('S channels) ('S depth))
+zipMatWith _ []   = Nothing
+zipMatWith f mats = Just . unsafePerformIO . exceptErrorIO $ do
+    resultMat <- newMat shape numChannels depth (toScalar (zero :: V4 CDouble))
+    lift $ withMatData resultMat $ \newStep newPtr ->
+        go (fromIntegral <$> newStep) (castPtr newPtr)
+    return $ unsafeCoerceMat resultMat
+  where
+    -- TODO: Compute the following from the statically known types
+    MatInfo shape !depth !numChannels = matInfo (head mats)
+
+    go :: Steps -> Ptr depth -> IO ()
+    go newStep newPtr =
+        forM_ (allPos shape) $ \pos ->
+            let pixels = getAllPixels (stepsAndPtrs mats) (fromIntegral numChannels) pos
+            in pokeArray (dest pos) (toList $ f pixels)
+      where
+        dest pos = matElemAddress (castPtr newPtr) newStep pos
+
+
+-- |Transforms a given list of matrices of equal shape, channels, and depth,
+-- by folding the given function over all matrix elements at each position.
+foldMat :: forall (shape :: [DS Nat]) (channels :: Nat) (depth :: *) a
+         . ( Storable depth, Storable a, All IsStatic shape )
+        => (a -> Pixel depth -> a)
+        -> a
+        -> [Mat ('S shape) ('S channels) ('S depth)]
+        -> Maybe (Vector a)
+foldMat _ _ []   = Nothing
+foldMat f z mats = Just . DV.fromList . map go $ allPos shape
+  where
+    go :: Position -> a
+    go = foldl' f z . getAllPixels (stepsAndPtrs mats) (fromIntegral numChannels)
+
+    -- TODO: Compute the following from the statically known types
+    MatInfo shape _ !numChannels = matInfo (head mats)
+
+
+stepsAndPtrs :: [Mat shape channels ('S depth)]
+             -> [(Steps, ForeignPtr depth)]
+stepsAndPtrs mats = unsafePerformIO $ forM mats $ \mat ->
+    withMatData mat $ \step ptr -> do
+        fptr <- newForeignPtr_ . castPtr $ ptr
+        return (fromIntegral <$> step, fptr)
+
+
+getAllPixels :: (Storable depth)
+             => [(Steps, ForeignPtr depth)]
+             -> Int
+             -> Position
+             -> [Pixel depth]
+getAllPixels stepsAndPtrs numChannels pos = map getAllPixels' stepsAndPtrs
+  where
+    getAllPixels' (step, dataPtr) =
+        let !offset = sum $ zipWith (*) step pos
+        in DV.unsafeFromForeignPtr dataPtr offset numChannels
```

Comment:
I see what you mean, but I need to have a second look as I still tend to think that it will work for both `zipMatWith` and `foldMat` because the pixels are eventually read from the unsafe Vector before the corresponding matrix (or matrices) can be garbage collected. At least in the case of `foldMat` there's the strict `foldl'` that happens during the mapping over all positions. Before the last position is reached the matrices cannot be garbage collected (but this could still be a problem for the very last position because the GC could garbage-collect between getting the unsafe vector from the last matrix and the `foldl'`). It's of course not a problem to change to a safer version where we copy the values but I wanted to avoid copying for performance. Just so that I understand better: I assume that there's a flaw in my reasoning above. Can you shed a light on this?